### ISSUE TITLE
websocket-polyfill headache

### ISSUE THREAD

**Issue by zappityzap** [nbd-wtf/nostr-tools#259]
TLDR: websocket-polyfill is junk and abandoned, can we switch nostr-tools to something better like isomorphic-ws?

I'm trying to use nostr-tools to add nostr DMs as a notification provider in Uptime Kuma, an open-source uptime monitoring tool. The PR was merged but then reverted because dependencies of websocket-polyfill don't offer binaries for arm.

websocket-polyfill depends on websocket (WebSocket-Node), which depends on bufferutil and utf-8-validate. bufferutil and utf-8-validate are subprojects of ws, and ws says they are optional dependencies for ws. websocket-polyfill uses websocket instead of ws.

bufferutil and utf-8-validate stopped providing arm binaries when something broke with cross-compiling. Since they are not optional dependencies for websocket, and arm binaries aren't available, they have to be built on arm.

Uptime Kuma insists on supporting arm v7 and not requiring users to build anything themselves, so binaries need to be available for all dependencies on all supported platforms.

websocket-polyfill hasn't been touched in 4 years and has almost no issues or PRs, open or closed. Cloning it to build locally doesn't work: it's missing a dev dependency for typescript. After installing that, tsc complains about type errors so it doesn't even build for me.

isomorphic-ws and ws are way more widely used then websocket-polyfill and websocket.

I replaced websocket-polyfill in nostr-tools with isomorphic-ws. It passes the nostr-tools tests and my Uptime Kuma PR works too. The only change to the code was one spot in relay.ts where I had to cast data to a string: `incomingMessageQueue.enqueue(e.data.toString('utf8'))`

I can try to get ws to start building arm binaries for bufferutil and utf-8-validate again, since they seem to be working now, but swapping out an old and busted websocket-polyfill for something getting widespread use and active development seems like a better fix.

**Comment by fiatjaf** [comment]
What do you think, @alexgleason?

**Comment by alexgleason** [comment]
Here are some examples of global Node.js polyfills:

- https://github.com/samchon/websocket-polyfill/blob/master/src/index.ts
- https://github.com/tgorka/node-window-polyfill/blob/master/register.js

Seems like you should be able to install `ws`, and then at some point early in your entrypoint of your application do this:

```ts
import ws from 'ws';

global.WebSocket = ws;
```

Boom, problem solved?

I wouldn't be against switching the devDependency to `isomorphic-ws` in this repo for tests, but I don't see the point in polyfilling it in the library itself, for the reasons I mentioned here: https://github.com/nbd-wtf/nostr-tools/pull/258#issuecomment-1646874210

When you choose Node.js in the modern world, you have chosen to have to think about polyfills, so in my view the polyfill should be added on the Node.js side.

**Comment by zappityzap** [comment]
Yes indeed, problem solved. This has really soured me on ever contributing to anything using node.
